File: backend/app/database/seed.py
import os
import json

from app.database.database import SessionLocal

from app.models.customer import Customer
from app.models.order import Order
from app.models.ticket import Ticket
from app.models.tool_action import ToolAction
from app.models.document import Document
from app.models.draft_reply import DraftReply
from app.models.triage_result import TriageResult

db = SessionLocal()

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))

DATA_DIR = os.path.join(BASE_DIR, "data")

def clear_database():
    db.query(Ticket).delete()
    db.query(Order).delete()
    db.query(Customer).delete()
    db.query(ToolAction).delete()
    db.query(Document).delete()
    db.query(DraftReply).delete()
    db.query(TriageResult).delete()

    db.commit()

    print("Database Cleared")

def load_json(filename):
    path = os.path.join(DATA_DIR, filename)

    with open(path, "r", encoding="utf-8") as file:
        return json.load(file)


def seed_customers():

    customers = load_json("customers.json")

    for customer in customers:

        db.add(

            Customer(

                customer_id=customer["customer_id"],

                name=customer["name"],

                email=customer["email"],

                tier=customer["tier"],

                country=customer["country"],

                created_at=customer["created_at"],

                verified=customer["verified"],

                tags=customer["tags"]

            )

        )

    db.commit()

    print(f"Inserted {len(customers)} customers")



def seed_orders():

    orders = load_json("orders.json")

    for order in orders:

        db.add(

            Order(

                order_id=order["order_id"],

                customer_id=order["customer_id"],

                status=order["status"],

                placed_at=order["placed_at"],

                delivered_at=order["delivered_at"],

                eligible_return_until=order["eligible_return_until"],

                total=order["total"],

                currency=order["currency"],

                payment_status=order["payment_status"],

                tracking_number=order["tracking_number"],

                items=order["items"]

            )

        )

    db.commit()

    print("Orders Seeded")



def seed_tickets():

    tickets = load_json("tickets.json")

    for ticket in tickets:

        db.add(

            Ticket(

                ticket_id=ticket["ticket_id"],

                customer_id=ticket["customer_id"],

                order_id=ticket["order_id"],

                channel=ticket["channel"],

                subject=ticket["subject"],

                body=ticket["body"],

                created_at=ticket["created_at"],

                status=ticket["status"]

            )

        )

    db.commit()

    print("Tickets Seeded")


def seed_tool_actions():

    actions = load_json("tool_actions.json")

    for action in actions:

        db.add(

            ToolAction(

                tool_name=action["tool_name"],

                description=action["description"],

                risk_level=action["risk_level"],

                requires_human_approval=action["requires_human_approval"],

                allowed_categories=action["allowed_categories"],

                required_fields=action["required_fields"]

            )

        )

    db.commit()

    print("Tool Actions Seeded")




def seed_documents():

    kb_path = os.path.join(DATA_DIR, "knowledge_base")

    for file in os.listdir(kb_path):

        if file.endswith(".md"):

            with open(

                os.path.join(kb_path, file),

                "r",

                encoding="utf-8"

            ) as f:

                content = f.read()

            db.add(

                Document(

                    document_id=file.replace(".md", ""),

                    title=file.replace(".md", "").replace("_", " ").title(),

                    category = file.replace(".md", ""),

                    content=content

                )

            )

    db.commit()

    print("Knowledge Base Seeded")        



if __name__ == "__main__":

    try:

        clear_database()

        seed_customers()
        seed_orders()
        seed_tickets()
        seed_tool_actions()
        seed_documents()

        print("Database Seeding Completed Successfully")

    except Exception as e:

        db.rollback()

        print(f"Error: {e}")

    finally:

        db.close()

File: backend/prisma/seed.js
const fs = require("fs/promises");
const path = require("path");
const { PrismaClient } = require("@prisma/client");

const prisma = new PrismaClient();
const dataDirectory = path.resolve(__dirname, "../../data");
const knowledgeBaseDirectory = path.join(dataDirectory, "knowledge_base");

async function readJson(filename) {
  const filePath = path.join(dataDirectory, filename);
  return JSON.parse(await fs.readFile(filePath, "utf8"));
}

function metadataValue(content, label) {
  const line = content
    .split(/\r?\n/)
    .find((value) => value.toLowerCase().startsWith(`${label.toLowerCase()}:`));

  if (!line) {
    throw new Error(`Knowledge document is missing '${label}:' metadata.`);
  }

  return line.split(":", 2)[1].trim();
}

function documentTitle(content, sourcePath) {
  const heading = content.split(/\r?\n/).find((line) => line.startsWith("# "));
  return heading ? heading.slice(2).trim() : path.basename(sourcePath, ".md");
}

async function readKnowledgeDocuments() {
  const filenames = (await fs.readdir(knowledgeBaseDirectory))
    .filter((filename) => filename.endsWith(".md"))
    .sort();

  return Promise.all(
    filenames.map(async (filename) => {
      const filePath = path.join(knowledgeBaseDirectory, filename);
      const content = await fs.readFile(filePath, "utf8");

      return {
        docId: metadataValue(content, "Doc ID"),
        title: documentTitle(content, filename),
        content,
        sourcePath: path.posix.join("data", "knowledge_base", filename),
        version: metadataValue(content, "Version"),
        audience: metadataValue(content, "Audience"),
      };
    }),
  );
}

async function seedCustomers(customers) {
  for (const customer of customers) {
    await prisma.customer.upsert({
      where: { customerId: customer.customer_id },
      create: {
        customerId: customer.customer_id,
        name: customer.name,
        email: customer.email,
        tier: customer.tier,
        country: customer.country,
        createdAt: new Date(customer.created_at),
        verified: customer.verified,
        tags: JSON.stringify(customer.tags),
      },
      update: {
        name: customer.name,
        email: customer.email,
        tier: customer.tier,
        country: customer.country,
        createdAt: new Date(customer.created_at),
        verified: customer.verified,
        tags: JSON.stringify(customer.tags),
      },
    });
  }
}

async function seedOrders(orders) {
  for (const order of orders) {
    await prisma.order.upsert({
      where: { orderId: order.order_id },
      create: {
        orderId: order.order_id,
        customerId: order.customer_id,
        status: order.status,
        placedAt: new Date(order.placed_at),
        deliveredAt: order.delivered_at ? new Date(order.delivered_at) : null,
        eligibleReturnUntil: order.eligible_return_until
          ? new Date(order.eligible_return_until)
          : null,
        total: order.total,
        currency: order.currency,
        paymentStatus: order.payment_status,
        trackingNumber: order.tracking_number ?? null,
        items: JSON.stringify(order.items),
      },
      update: {
        customerId: order.customer_id,
        status: order.status,
        placedAt: new Date(order.placed_at),
        deliveredAt: order.delivered_at ? new Date(order.delivered_at) : null,
        eligibleReturnUntil: order.eligible_return_until
          ? new Date(order.eligible_return_until)
          : null,
        total: order.total,
        currency: order.currency,
        paymentStatus: order.payment_status,
        trackingNumber: order.tracking_number ?? null,
        items: JSON.stringify(order.items),
      },
    });
  }
}

async function seedTickets(tickets) {
  for (const ticket of tickets) {
    await prisma.ticket.upsert({
      where: { ticketId: ticket.ticket_id },
      create: {
        ticketId: ticket.ticket_id,
        customerId: ticket.customer_id,
        orderId: ticket.order_id ?? null,
        channel: ticket.channel,
        subject: ticket.subject,
        body: ticket.body,
        createdAt: new Date(ticket.created_at),
        status: ticket.status,
        expectedCategory: ticket.expected_category ?? null,
        expectedPriority: ticket.expected_priority ?? null,
        expectedSentiment: ticket.expected_sentiment ?? null,
        expectedEscalation: ticket.expected_escalation ?? null,
        expectedActions: JSON.stringify(ticket.expected_actions ?? []),
      },
      update: {
        customerId: ticket.customer_id,
        orderId: ticket.order_id ?? null,
        channel: ticket.channel,
        subject: ticket.subject,
        body: ticket.body,
        createdAt: new Date(ticket.created_at),
        status: ticket.status,
        expectedCategory: ticket.expected_category ?? null,
        expectedPriority: ticket.expected_priority ?? null,
        expectedSentiment: ticket.expected_sentiment ?? null,
        expectedEscalation: ticket.expected_escalation ?? null,
        expectedActions: JSON.stringify(ticket.expected_actions ?? []),
      },
    });
  }
}

async function seedKnowledgeDocuments(documents) {
  for (const document of documents) {
    await prisma.knowledgeDocument.upsert({
      where: { docId: document.docId },
      create: document,
      update: document,
    });
  }
}

function ticketForTool(tool, tickets) {
  const ticket = tickets.find((candidate) =>
    tool.allowed_categories.includes(candidate.expected_category),
  );

  if (!ticket) {
    throw new Error(`No seed ticket is compatible with tool '${tool.tool_name}'.`);
  }

  return ticket.ticket_id;
}

async function seedToolActions(tools, tickets) {
  for (const tool of tools) {
    const idempotencyKey = `catalog:${tool.tool_name}`;
    const data = {
      ticketId: ticketForTool(tool, tickets),
      toolName: tool.tool_name,
      payload: JSON.stringify({
        description: tool.description,
        allowedCategories: tool.allowed_categories,
        requiredFields: tool.required_fields,
        maxAmountInr: tool.max_amount_inr ?? null,
      }),
      riskLevel: tool.risk_level,
      requiresHumanApproval: tool.requires_human_approval,
      status: "catalog",
      idempotencyKey,
    };

    await prisma.toolAction.upsert({
      where: { idempotencyKey },
      create: data,
      update: data,
    });
  }
}

async function main() {
  const [customers, orders, tickets, tools, knowledgeDocuments] = await Promise.all([
    readJson("customers.json"),
    readJson("orders.json"),
    readJson("tickets.json"),
    readJson("tool_actions.json"),
    readKnowledgeDocuments(),
  ]);

  await seedCustomers(customers);
  await seedOrders(orders);
  await seedTickets(tickets);
  await seedKnowledgeDocuments(knowledgeDocuments);
  await seedToolActions(tools, tickets);

  console.log("TrustDesk seed completed.");
}

main()
  .catch((error) => {
    console.error("TrustDesk seed failed:", error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
